draft of benchmark report (#270)
megaease/easeagent: doc/benchmark.md

     http_req_duration..............: avg=360.83ms min=1.45ms med=252.27ms max=7.99s    p(90)=737.55ms p(95)=1.04s

       { expected_response:true }...: avg=360.83ms min=1.45ms med=252.27ms max=7.99s    p(90)=737.55ms p(95)=1.04s
     http_req_failed................: 0.00%   ✓ 0           ✗ 1222680
     http_req_receiving.............: avg=4.88ms   min=9.14µs med=51.6µs   max=887.76ms p(90)=10.7ms   p(95)=28.04ms
     http_req_sending...............: avg=193.63µs min=3.72µs med=8.54µs   max=756.82ms p(90)=19.57µs  p(95)=44.94µs
     http_req_tls_handshaking.......: avg=0s       min=0s     med=0s       max=0s       p(90)=0s       p(95)=0s
     http_req_waiting...............: avg=355.75ms min=1.36ms med=248.41ms max=7.99s    p(90)=727.94ms p(95)=1.03s

     http_reqs......................: 1222680 1526.574647/s

     iteration_duration.............: avg=1.62s    min=1s     med=1.41s    max=8.99s    p(90)=2.36s    p(95)=2.76s
     iterations.....................: 305670  381.643662/s
     vus............................: 6       min=5         max=700
     vus_max........................: 700     min=700       max=700
```

## Conclusion

> Optimisation is required in terms of cpu efficiency/memory usage and latency throughput.

Reviewer: ```suggestion
**Optimization is required in terms of CPU efficiency/memory usage and latency throughput.** 
```

> The average heap memory has been growing slowly, so there may be a memory leak that needs to be addressed.

Author: OK, I check to draw a clear conclusion latter

# Benchmark
We use [spring-petclinic applications](https://github.com/spring-petclinic/spring-petclinic-microservices) as the benchmark application because it is relatively close to the real user scenario.

We use three AWS EC2 machines in a subnet for the benchmark test:

- Baseline EC2: run the application stack without Easeagent.
- Agent EC2: run the application with Easeagent.
- Tester EC2: we initiate testing requests and collect testing metric data on this node.

We run Jmeter on `Tester` node to start two identical thread groups to access the `Baseline` node and `Agent` node at the same time, and collect the metric data.

## Test Result

> We access two URIs of the application.

Reviewer: ```suggestion
The following two URIs which we use to test
```

- API-URI-owner-3spans: /api/gateway/owners/2
- API-URI-vets-10spans: /api/vet/vets

> These two URIs generate 3 and 10 SPANs respectively, and by comparing the results of the two data the impact of the number of SPANs can be observed.

Reviewer: ```suggestion
These two URIs generate 3 and 10 SPANs respectively, and by comparing the resource utilization metrics,  the impact of the number of SPANs can be observed.
```

For each value pair in the table, the former is the baseline value and the latter is the agent value (baseline / agent / difference value).

> - 10 threads groups 

Reviewer: ```suggestion
### 10 threads groups 
```

| URI-Label                | CPU (baseline/agent/diff) | Heap Memory        | P75 Latency               | Throughput (15s) |
|----------------------|:--------------------------|:-------------------|:--------------------------|------------------|
| API-URI-owner-3spans | 10.7% / 13.3% / 2.5%      | 246M / 289M / 43M  | 100ms  / 120ms  / 20ms    | 361 / 293 / -68  |
| API-URI-vets-10spans | 8.56% / 11.3% / 2.64%     | 261M / 200M / -61M | 58.8ms / 88.8ms / 30ms    | 361 / 293 / -68  |

> - 20 threads groups 

Reviewer: ```suggestion
### 20 threads groups 
```

| URI-Label                | CPU (baseline/agent/diff) | Heap Memory        | P75 Latency               | Throughput (15s) |
|----------------------|:--------------------------|:-------------------|:--------------------------|------------------|
| API-URI-owner-3spans | 10.0% / 12.8% / 2.8%      | 244M / 282M / 38M  | 43ms / 49ms / 6ms         | 332 / 277 / -55  |
| API-URI-vets-10spans | 8.43% / 11.0% / 2.57%     | 179M / 255M / -71M | 26ms / 36ms / 10ms        | 332 / 277 / -55  |

> - 50 threads groups 

Reviewer: ```suggestion
### 50 threads groups 
```

| URI-Label                | CPU (baseline/agent/diff) | Heap Memory        | P75 Latency               | Throughput (15s) |
|----------------------|:--------------------------|:-------------------|:--------------------------|------------------|
| API-URI-owner-3spans | 10.3% / 12.8% / 2.5%      | 263M / 296M / 33M  | 214ms / 241ms / 27ms      | 357 / 300 / -57   |
| API-URI-vets-10spans | 8.12% / 10.5% / 2.38%     | 175M / 191M / 16M  | 117ms / 161ms / 44ms      | 357 / 300 / -57   |

> - Process CPU Load: Easeagent will increase the process CPU load by about 3%.

Reviewer: Move to conclusion

- Heap Memory: EaseAgent has a minimal impact on memory usage.
- Latency & Throughput: Easeagent has an excessive impact on latency and throughput and is positively correlated with the number of Spans.

CPU, Process CPU Load, collected through JMX:java_lang_OperatingSystem_ProcessCpuLoad, this may be replaced by other method.

> ### Data of 10 Threads Groups 

Reviewer: ```suggestion
### Diagram of 10 Threads Groups 
```

> ### Data of 10 Threads Groups 

Reviewer: Move to the position that is following the 10 thread result table.

#### CPU
![Process CPU Load](./images/benchmark/10t-process-cpu-load-mean.png)

#### Heap
![Process CPU Load](./images/benchmark/10t-mem-heap.png)

#### Lagency & Throuthput
![Process CPU Load](./images/benchmark/10t-resp-time-p75.png)
![Process CPU Load](./images/benchmark/10t-resp-per-15seconds.png)

> ### Data of 20 Threads Groups 

Reviewer: ```suggestion
### Diagram of 20 Threads Groups 
```

> ### Data of 20 Threads Groups 

Reviewer: Move to the position that is following the 20 thread result table.

#### CPU
![Process CPU Load](./images/benchmark/20t-process-cpu-load-mean.png)

#### Heap
![Process CPU Load](./images/benchmark/20t-mem-heap.png)

#### Lagency & Throuthput
![Process CPU Load](./images/benchmark/20t-resp-time-p75.png)
![Process CPU Load](./images/benchmark/20t-resp-per-15seconds.png)

> ### Data of 50 Threads Groups 

Reviewer: ```suggestion
### Diagram of 50 Threads Groups 
```

> ### Data of 50 Threads Groups 

Reviewer: Move to the position that is following the 50 thread result table.
